portico: Keep targeted elements below menu bar.

karlstolley opened the pull request:
This PR ensure that all elements targeted by URL fragments will remain visible below the portico's menu bar at all viewport sizes and also when a user zooms in.

The PR also includes adjustments to the shared structure across similar portico landing pages by introducing an `inner-content` div on the few pages missing one, and it also orrects some incidental missing heading IDs.

Thanks to @andersk for suggesting the use of `scroll-margin-top`.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

Before:
<img width="1115" alt="fragment-before" src="https://github.com/zulip/zulip/assets/170719/33fd44cc-c8cb-42f7-83ac-486d4d920a54">

With this PR:
<img width="1115" alt="fragment-after" src="https://github.com/zulip/zulip/assets/170719/640e9873-9cc3-4660-8c05-b0ba011f1a6d">

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

andersk commented on web/styles/portico/portico.css:
```diff
@@ -16,6 +16,12 @@ html {
     position: fixed;
 }
 
+:target {
+    /* Keep elements targeted by URL fragments visible below the
+    floating top menu; coordinate the value here with the menu's height. */
+    scroll-margin-top: 60px;
+}
```
This selector is too broad. We apply `portico.css` on basically every page that isn’t the main app, including the login, billing, and help pages, which have a variety of different scrolling setups.

karlstolley replied:
Got it. So it seems like no only should the selector be more targeted, but maybe it belongs in the `landing_page.css` file, too?

Since the original for this referenced case-study pages, maybe `.case-study-page :target` is the selector to start with. But I'll ask [over on CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/landing.20page.20anchor.20links/near/1566217).

timabbott replied:
I think there's two possible changes here. One idea would be to add this selector in a CSS file that's included precisely for the pages where this is needed; but I'm not sure how to compute that easily. 

I think `.portico-landing` might be the right class for "page with landing page navigation".

```
$ git grep  --files-with-match 'class=.*portico-landing'  > /tmp/a
$ git grep  --files-with-match "include 'zerver/landing_nav.html" >/tmp/b
$ diff -ur /tmp/a /tmp/b 
--- /tmp/a	2023-05-10 12:08:31.825913529 -0700
+++ /tmp/b	2023-05-10 12:07:43.100974445 -0700
@@ -1,6 +1,5 @@
 templates/corporate/apps.html
 templates/corporate/attribution.html
-templates/corporate/billing.html
 templates/corporate/case-studies/asciidoctor-case-study.html
 templates/corporate/case-studies/atolio-case-study.html
 templates/corporate/case-studies/end-point-case-study.html
@@ -12,7 +11,6 @@
 templates/corporate/case-studies/ucsd-case-study.html
 templates/corporate/communities.html
 templates/corporate/development-community.html
-templates/corporate/event_status.html
 templates/corporate/features.html
 templates/corporate/for/business.html
 templates/corporate/for/communities.html
@@ -29,7 +27,6 @@
 templates/corporate/self-hosting.html
 templates/corporate/team.html
 templates/corporate/try-zulip.html
-templates/corporate/upgrade.html
 templates/corporate/values.html
 templates/corporate/why-zulip.html
 templates/corporate/zephyr.html
```

The three pages in the diff are billing system pages that I would expect to not have the `portico-landing` class at all, I'll change that, but then after that I think this is the correct rule.

timabbott replied:
https://github.com/zulip/zulip/pull/25518 fixes that issue; following it the above diff is the empty set.

karlstolley commented:
Thanks, Anders and Tim. I set up the new selector in the `landing_page.css` file, so hopefully this is ready for review and merge.

timabbott commented:
Code looks good to me conceptually; @alya maybe you should do a round of testing. These are the affected pages:
```
$ git grep portico-landing templates/
templates/corporate/apps.html:<div class="portico-landing apps">
templates/corporate/attribution.html:<div class="portico-landing why-page">
templates/corporate/case-studies/asciidoctor-case-study.html:<div class="portico-landing why-page solutions-page case-study-page">
templates/corporate/case-studies/atolio-case-study.html:<div class="portico-landing why-page solutions-page case-study-page">
templates/corporate/case-studies/end-point-case-study.html:<div class="portico-landing why-page solutions-page case-study-page">
templates/corporate/case-studies/idrift-case-study.html:<div class="portico-landing why-page solutions-page case-study-page">
templates/corporate/case-studies/lean-case-study.html:<div class="portico-landing why-page solutions-page case-study-page">
templates/corporate/case-studies/recurse-center-case-study.html:<div class="portico-landing why-page solutions-page case-study-page">
templates/corporate/case-studies/rust-case-study.html:<div class="portico-landing why-page solutions-page case-study-page">
templates/corporate/case-studies/tum-case-study.html:<div class="portico-landing why-page solutions-page case-study-page">
templates/corporate/case-studies/ucsd-case-study.html:<div class="portico-landing why-page solutions-page case-study-page">
templates/corporate/communities.html:<div class="portico-landing integrations communities">
templates/corporate/communities.html:</div> <!-- .portico-landing -->
templates/corporate/development-community.html:<div class="portico-landing why-page solutions-page">
templates/corporate/features.html:<div class="portico-landing features-app">
templates/corporate/for/business.html:<div class="portico-landing plans why-page solutions-page for-companies">
templates/corporate/for/communities.html:<div class="portico-landing why-page solutions-page">
templates/corporate/for/education.html:<div class="portico-landing why-page solutions-page">
templates/corporate/for/events.html:<div class="portico-landing why-page solutions-page">
templates/corporate/for/open-source.html:<div class="portico-landing why-page solutions-page">
templates/corporate/for/research.html:<div class="portico-landing why-page solutions-page">
templates/corporate/for/use-cases.html:<div class="portico-landing plans why-page solutions-page for-companies">
templates/corporate/hello.html:<div class="portico-landing hello show">
templates/corporate/history.html:<div class="portico-landing why-page solutions-page">
templates/corporate/jobs.html:<div class="portico-landing jobs why-page">
templates/corporate/plans.html:<div class="portico-landing plans">
templates/corporate/security.html:<div class="portico-landing why-page">
templates/corporate/self-hosting.html:<div class="portico-landing why-page solutions-page self-hosting-page">
templates/corporate/team.html:<div class="portico-landing why-page">
templates/corporate/try-zulip.html:<div class="portico-landing why-page solutions-page try-zulip-now-page">
templates/corporate/values.html:<div class="portico-landing why-page solutions-page">
templates/corporate/why-zulip.html:<div class="portico-landing why-page">
templates/corporate/zephyr.html:<div class="portico-landing why-page">
templates/zerver/integrations/index.html:<div class="portico-landing integrations">
templates/zerver/integrations/index.html:</div> <!-- .portico-landing -->
```

alya commented:
I spot-checked a bunch, and some were OK, while others didn't quite work right.

On /why-zulip:
<img width="789" alt="Screen Shot 2023-05-10 at 9 49 49 PM" src="https://github.com/zulip/zulip/assets/2090066/4322b18a-106c-48ff-a336-827b093fe273">

Here's another one where the first heading doesn't look right: http://localhost:9991/values/#building-software-that-will-always-be-there-for-our-users
<img width="799" alt="Screen Shot 2023-05-10 at 9 53 36 PM" src="https://github.com/zulip/zulip/assets/2090066/f79f708c-72d2-4ea2-a0ed-7abcc784152d">

The ones on /team end up a bit higher than I'd expect:
<img width="654" alt="Screen Shot 2023-05-10 at 9 51 35 PM" src="https://github.com/zulip/zulip/assets/2090066/fa67fd85-c8e3-4112-9ae9-5515f622f7b1">
<img width="646" alt="Screen Shot 2023-05-10 at 9 51 29 PM" src="https://github.com/zulip/zulip/assets/2090066/cf26a6b9-af45-43ba-9690-99f2b9768e3a">

alya commented:
That said, it definitely seems like an improvement over `main`.

alya commented:
As a side note, why does the "Open positions" heading on /jobs go to /jobs/#undefined?

karlstolley commented:
Hi @alya -- Fixes incoming, but these appear to be the issues:

* The `h1` styles on `/team` have some mystery-meat `::before` selectors that are gumming up where the `h1` content begins, which is causing Chrome to appear to show text from the paragraph above the `h1`
* The hidden/half-visible heading issues on `/why-zulip` and `/values` appear to be caused by different approaches to the hero div on those pages.
* "Open positions" is missing an ID in the template HTML file, but it's a quick fix to add `id="open-positions"` there.

Interestingly, none of the scroll issues present in Firefox. So these appear to be related to how Chrome calculates the heights and positions of things. But I think I can fix it all the same.

alya commented:
I'd be fine with leaving muti-line headings as an open issue.

I just tested http://localhost:9991/use-cases/#use-cases with the current state of this PR, and it wasn't working right either, so I think there's more to fix here?

<img width="778" alt="Screen Shot 2023-05-11 at 2 13 18 PM" src="https://github.com/zulip/zulip/assets/2090066/fe80a2db-9515-404e-b596-33d7ae95719d">

karlstolley commented:
You're absolutely right, @alya!

I spent my morning rethinking my approach, and I have good reason to believe that this is now fixed in Chrome as well as Firefox, and correctly handles even multiline headings.

The catch was that outer padding was not being figured into `scroll-margin-top` on Chrome, which is why headings at the very top of the page were never rendering correctly. It seems, too, that that issue was causing the problem with multiline headings, as that issue disappeared after I introduced the fix originally for the initial headings.

Please give this another test, if you have the time.

alya commented:
Everything I tested is looking good!

timabbott commented:
I pushed back a bonus commit that mostly adds a new CSS variable, but I also moved around the comments a bit.

In general as a project we prefer the style of comments on the line before what they're explaining, rather than at the end of lines; I find inline comments a bit harder to read and also often terser than they would ideally be due to the cramped context.

karlstolley commented on web/styles/portico/landing_page.css:
```diff
@@ -1,3 +1,7 @@
+:root {
+    --padded-content-padding: 50px;
```
Nice! That makes good sense.
